snappy/security.go

```diff
@@ -121,106 +329,616 @@ func getSecurityProfile(m *packageYaml, appName, baseDir string) (string, error)
+func CompareGeneratePolicyFromFile(fn string) error {
+	m, err := parsePackageYamlFileWithVersion(fn)
 	if err != nil {
-		logger.Noticef("%v failed", args)
+		return err
 	}
 
-	return content, err
+	baseDir := filepath.Dir(filepath.Dir(fn))
+
+	for _, service := range m.ServiceYamls {
+		p, err := service.generatePolicyForServiceBinaryResult(m, service.Name, baseDir)
+
+		// FIXME: use apparmor_profile -p on both AppArmor profiles
+
+		if err != nil {
+			// FIXME: what to do here?
+			return err
+		}
+		if err := comparePolicyToCurrent(p); err != nil {
+			return err
+		}
+	}
+
+	for _, binary := range m.Binaries {
+		p, err := binary.generatePolicyForServiceBinaryResult(m, binary.Name, baseDir)
+		if err != nil {
+			// FIXME: what to do here?
+			return err
+		}
+		if err := comparePolicyToCurrent(p); err != nil {
+			return err
+		}
+	}
+
+	// now compare the snappy-config profile
+	if hasConfig(baseDir) {
+		p, err := snappyConfig.generatePolicyForServiceBinaryResult(m, "snappy-config", baseDir)
+		if err != nil {
+			return nil
+		}
+		if err := comparePolicyToCurrent(p); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// FIXME: refactor so that we don't need this
+func parsePackageYamlFileWithVersion(fn string) (*packageYaml, error) {
+	m, err := parsePackageYamlFile(fn)
+
+	// FIXME: duplicated code from snapp.go:NewSnapPartFromYaml,
+	//        version is overriden by sideloaded versions
+	m.Version = filepath.Base(filepath.Dir(filepath.Dir(fn)))
+
+	return m, err
 }
 
-func readSeccompOverride(yamlPath string, s *securitySeccompOverride) error {
-	yamlData, err := ioutil.ReadFile(yamlPath)
+// GeneratePolicyFromFile is used to generate security policy on the system
+// from the specified manifest file name
+func GeneratePolicyFromFile(fn string, force bool) error {
+	// FIXME: force not used yet
+	m, err := parsePackageYamlFileWithVersion(fn)
+	if err != nil {
+		return err
+	}
+
+	if m.Type == "" || m.Type == pkg.TypeApp {
+		_, err = originFromYamlPath(fn)
+		if err != nil {
+			if err == ErrInvalidPart {
+				err = errOriginNotFound
+			}
+			return err
+		}
+	}
+
+	// TODO: verify cache files here
+
+	baseDir := filepath.Dir(filepath.Dir(fn))
+	err = generatePolicy(m, baseDir)
 	if err != nil {
 		return err
 	}
 
-	err = yaml.Unmarshal(yamlData, &s)
+	return err
+}
+
+// RegenerateAllPolicy will re-generate all policy that needs re-generating
+func RegenerateAllPolicy(force bool) error {
+	installed, err := NewMetaLocalRepository().Installed()
 	if err != nil {
-		return &ErrInvalidYaml{File: "package.yaml[seccomp override]", Err: err, Yaml: yamlData}
+		return err
 	}
-	// These must always be specified together
-	if (s.PolicyVersion == 0 && s.PolicyVendor != "") || (s.PolicyVersion != 0 && s.PolicyVendor == "") {
-		return ErrInvalidSeccompPolicy
+
+	for _, p := range installed {
+		if _, ok := p.(*SnapPart); !ok {
+			continue
+		}
+		basedir := p.(*SnapPart).basedir
```

Comment:
Thanks! I fixed that double type assertion. And we have a "part.Dir()" in the squashfs branch iirc (and if not we will add it :)